torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -125,9 +132,81 @@ class ProcessGroupNCCL : public ProcessGroup {
     // to the store.
     std::shared_ptr<Store> store_;
 
+    // Store a Future work associated with WorkNCCL.
+    c10::intrusive_ptr<c10::ivalue::Future> futureWork_;
+
     friend class ProcessGroupNCCL;
   };
 
+  // FutureNCCL is a subclass of c10d ivalue's Future. The goal is to
+  // use this class in getFuture API of WorkNCCL. This Future is mostly a
+  // wrapper to synchronize streams appropriately and it mostly enables the
+  // async programming model of CUDA while trying to adhere to the Future
+  // interface.
+  //
+  // FutureNCCL has a reference to WorkNCCL and NCCL collective's outputs.
+  // Its value is NCCL collective's outputs and should only be used after
+  // NCCL kernel is done executing.
+  struct FutureNCCL : at::ivalue::Future {
+   public:
+    explicit FutureNCCL(
+        std::shared_ptr<ProcessGroupNCCL::WorkNCCL> work,
+        std::vector<at::Tensor>& outputs)
+        : at::ivalue::Future(c10::ListType::create(c10::TensorType::get())),
+          work_(work),
+          outputs_(outputs) {}
+
+    // Simply calls WorkNCCL's wait(). It will return after
+    // synchronizing the correct GPU streams to ensure we can have async CUDA
+    // execution and it does not wait for the entire operation to complete on
+    // GPU. If NCCL_BLOCKING_WAIT is enabled, in that case, it will wait for the
+    // entire operation to complete before returning.
+    void wait() override {
+      work_->wait();
+    }
+
+    // FutureNCCL's value is NCCL collective's outputs and callbacks were
+    // invoked inline by addCallback(), so markCompleted is not needed.
+    void markCompleted(at::IValue value) override {
+      TORCH_CHECK(false, "FutureNCCL::markCompleted is not supported.");
+    }
+
+    // Returns NCCL collective's outputs. This should only be used after NCCL
+    // kernel is done executing.
+    at::IValue value() override {
+      return outputs_;
+    }
+
+    const at::IValue& constValue() override {
+      return outputs_;
+    }
```

Comment:
We probably need to call `work_->wait()` here as well to ensure we synchronize the streams correctly.

Reply:
After this change, I guess we don't need `work_->wait()` in `addCallback()` anymore:
https://github.com/pytorch/pytorch/blob/7352d01a980a8d724c061f2bfc0bc1fa4b164ac5/torch/lib/c10d/ProcessGroupNCCL.hpp#L187-L191

Reply:
I guess @pritamdamania87 is referring to `value()` and `constValue()`?

Reply:
Right, but if we call `work_->wait()` inside `value`, there is no need to call it before invoking the callback inline inside `addCallback()`.

Reply:
is that some newer version that hasn't been published yet? Looks like `work_->wait()` is not in `value()` in the current version?

Reply:
I'll update it soon.